### ISSUE TITLE
undesirable_function_linter() could support a custom regex?

### ISSUE THREAD

**Issue by MichaelChirico** [r-lib/lintr#2701]
I am looking at the default `undesirable_function_linter()` throwing for this:

```r
par('usr')
```

because the default looks for any call to `par()`. But actually, what the lint reasoning really cares about is _writing_ device settings, not _reading_ them.

So we might want to tweak the XPath for `par()`: `//SYMBOL_FUNCTION_CALL[text() = 'par']/parent::expr/following-sibling::EQ_SUB`, i.e. it has to use a keyword argument.

I think this feature is quite generalizable -- I've definitely encountered situations where a function is undesirable in some-but-not-all circumstances, and rather than introducing a new `par_linter()` (or generally `foo_linter()`) every time there's a minor rule that's just outside of what `undesirable_function_linter()` can accomplish, it would be good to extend `undesirable_function_linter()` itself to be able to handle these situtations.

What remains is to come up with a good design for this. The first thing that comes to mind is to accept a list instead of just a string:

```r
undesirable_function_linter(c(par = '...'))

undesirable_function_linter(list(par = list(lint_message = '...', xpath = '...')))
```

I think this would also help with (if not obviate) #1912, #1770.
